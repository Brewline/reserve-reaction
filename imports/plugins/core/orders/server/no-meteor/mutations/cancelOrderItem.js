import SimpleSchema from "simpl-schema";
import ReactionError from "@reactioncommerce/reaction-error";
import Random from "@reactioncommerce/random";
import { Order as OrderSchema } from "/imports/collections/schemas";
import updateGroupStatusFromItemStatus from "../util/updateGroupStatusFromItemStatus";

const canceledStatus = "coreOrderWorkflow/canceled";
const itemCanceledStatus = "coreOrderItemWorkflow/canceled";

// These should eventually be configurable in settings
const itemStatusesThatOrdererCanCancel = ["new"];
const orderStatusesThatOrdererCanCancel = ["new"];

const inputSchema = new SimpleSchema({
  cancelQuantity: {
    type: SimpleSchema.Integer,
    min: 1
  },
  itemId: String,
  orderId: String,
  reason: {
    type: String,
    optional: true
  }
});

/**
 * @method cancelOrderItem
 * @summary Use this mutation to cancel one item of an order, either for the
 *   full ordered quantity or for a partial quantity. If partial, the item will be
 *   split into two items and the original item will have a lower quantity and will
 *   be canceled.
 *
 *   If this results in all items in a fulfillment group being canceled, the group
 *   will also be canceled. If this results in all fulfillment groups being canceled,
 *   the full order will also be canceled.
 * @param {Object} context - an object containing the per-request state
 * @param {Object} input - Necessary input. See SimpleSchema
 * @return {Promise<Object>} Object with `order` property containing the created order
 */
export default async function cancelOrderItem(context, input) {
  inputSchema.validate(input);

  const {
    orderId,
    itemId,
    cancelQuantity,
    reason = null
  } = input;

  const { accountId, appEvents, collections, isInternalCall, userHasPermission, userId } = context;
  const { Orders } = collections;

  // First verify that this order actually exists
  const order = await Orders.findOne({ _id: orderId });
  if (!order) throw new ReactionError("not-found", "Order not found");

  // Allow cancel if the account that placed the order is attempting to cancel
  // or if the account has "orders" permission. When called internally by another
  // plugin, context.isInternalCall can be set to `true` to disable this check.
  if (
    !isInternalCall &&
    (!accountId || accountId !== order.accountId) &&
    !userHasPermission(["orders"], order.shopId)
  ) {
    throw new ReactionError("access-denied", "Access Denied");
  }

  // Is the account calling this mutation also the account that placed the order?
  // We need this check in a couple places below, so we'll get it here.
  const accountIsOrderer = (order.accountId && accountId === order.accountId);

  // The orderer may only cancel while the order status is still "new"
  if (accountIsOrderer && !orderStatusesThatOrdererCanCancel.includes(order.workflow.status)) {
    throw new ReactionError("invalid", `Order status (${order.workflow.status}) is not one of: ${orderStatusesThatOrdererCanCancel.join(", ")}`);
  }

  // Find and cancel the item
  let foundItem = false;
  const updatedGroups = order.shipping.map((group) => {
    let itemToAdd;
    const updatedItems = group.items.map((item) => {
      if (item._id !== itemId) return item;
      foundItem = true;

      // The orderer may only cancel while the order item status is still "new"
      if (accountIsOrderer && !itemStatusesThatOrdererCanCancel.includes(item.workflow.status)) {
        throw new ReactionError("invalid", `Item status (${item.workflow.status}) is not one of: ${itemStatusesThatOrdererCanCancel.join(", ")}`);
      }

      // If they are requesting to cancel fewer than the total quantity of items
      // that were ordered, we'll create a new item here with the remaining quantity.
      // It will have the same status as this item has before we cancel it.
      // Later, after we exit this loop, we'll append it to the `group.items` list.
      if (item.quantity > cancelQuantity) {
        const newItemQuantity = item.quantity - cancelQuantity;
        itemToAdd = {
          ...item,
          _id: Random.id(),
          quantity: newItemQuantity
        };

        // Update the subtotal since it is related to the quantity
        itemToAdd.subtotal = item.price.amount * newItemQuantity;
      } else if (item.quantity < cancelQuantity) {
        throw new ReactionError("invalid-param", "cancelQuantity may not be greater than item quantity");
      }

      const updatedItem = {
        ...item,
        cancelReason: reason,
        quantity: cancelQuantity
      };

      // Update the subtotal since it is related to the quantity
      updatedItem.subtotal = item.price.amount * cancelQuantity;

      if (item.workflow.status !== itemCanceledStatus) {
        updatedItem.workflow = {
          status: itemCanceledStatus,
          workflow: [...item.workflow.workflow, itemCanceledStatus]
        };
      }

      // If we make it this far, then we've found the item that they want to cancel.
      // We set the status and the cancel reason if one was provided.
      // This will also decrement the quantity to match the quantity that is being
      // canceled, which will be offset by pushing `itemToAdd` into the array later.
      return updatedItem;
    });

    // If they canceled fewer than the full quantity of the item, add a new
    // non-canceled item to make up the difference.
    if (itemToAdd) {
      updatedItems.push(itemToAdd);
    }

    const updatedGroup = { ...group, items: updatedItems };

    // Ensure proper group status
    updateGroupStatusFromItemStatus(updatedGroup);

    // There is a convenience itemIds prop, so update that, too
    if (itemToAdd) {
      updatedGroup.itemIds.push(itemToAdd._id);
    }

    // Return the group, with items and workflow potentially updated.
    return updatedGroup;
  });

  // If we did not find any matching item ID while looping, something is wrong
  if (!foundItem) throw new ReactionError("not-found", "Order item not found");

  // If all groups are canceled, set the order status to canceled
  let updatedOrderWorkflow;
  let fullOrderWasCanceled = false;
  const allGroupsAreCanceled = updatedGroups.every((group) => group.workflow.status === canceledStatus);
  if (allGroupsAreCanceled && order.workflow.status !== canceledStatus) {
    updatedOrderWorkflow = {
      status: canceledStatus,
      workflow: [...order.workflow.workflow, canceledStatus]
    };
    fullOrderWasCanceled = true;
  }

  // We're now ready to actually update the database and emit events
  const modifier = {
    $set: {
      shipping: updatedGroups,
      updatedAt: new Date()
    }
  };

  if (updatedOrderWorkflow) {
    modifier.$set.workflow = updatedOrderWorkflow;
  }

  OrderSchema.validate(modifier, { modifier: true });

  const { modifiedCount, value: updatedOrder } = await Orders.findOneAndUpdate(
    { _id: orderId },
    modifier,
    { returnOriginal: false }
  );
  if (modifiedCount === 0 || !updatedOrder) throw new ReactionError("server-error", "Unable to update order");

  await appEvents.emit("afterOrderUpdate", {
    order: updatedOrder,
    updatedBy: userId
  });

  if (fullOrderWasCanceled) {
    await appEvents.emit("afterOrderCancel", {
      canceledBy: userId,
      order: updatedOrder
    });
  }

  return { order: updatedOrder };
}
