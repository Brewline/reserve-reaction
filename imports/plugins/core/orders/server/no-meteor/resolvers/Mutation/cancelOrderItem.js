import { decodeOrderOpaqueId, decodeOrderItemOpaqueId } from "@reactioncommerce/reaction-graphql-xforms/order";

/**
 * @name Mutation/cancelOrderItem
 * @method
 * @memberof Payments/GraphQL
 * @summary resolver for the cancelOrderItem GraphQL mutation
 * @param {Object} parentResult - unused
 * @param {Object} args.input - an object of all mutation arguments that were sent by the client
 * @param {String} args.input.orderId - The order ID
 * @param {String} args.input.itemId - The ID of the item to cancel
 * @param {Number} args.input.cancelQuantity - Quantity to cancel. Must be equal to or less than the item quantity.
 * @param {String} [args.input.reason] - Optional free text reason for cancel
 * @param {String} [args.input.clientMutationId] - An optional string identifying the mutation call
 * @param {Object} context - an object containing the per-request state
 * @return {Promise<Object>} CancelOrderItemPayload
 */
export default async function cancelOrderItem(parentResult, { input }, context) {
  const {
    clientMutationId = null,
    cancelQuantity,
    itemId,
    orderId,
    reason
  } = input;

  const { order } = await context.mutations.cancelOrderItem(context, {
    cancelQuantity,
    itemId: decodeOrderItemOpaqueId(itemId),
    orderId: decodeOrderOpaqueId(orderId),
    reason
  });

  return {
    clientMutationId,
    order
  };
}
