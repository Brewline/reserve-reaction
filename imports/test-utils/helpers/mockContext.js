const mockContext = {
  accountId: "FAKE_ACCOUNT_ID",
  appEvents: {
    emit() {},
    on() {}
  },
  collections: {},
  getFunctionsOfType: jest.fn().mockName("getFunctionsOfType").mockReturnValue([]),
  shopId: "FAKE_SHOP_ID",
  userHasPermission: jest.fn().mockName("userHasPermission"),
  userId: "FAKE_USER_ID"
};

[
  "Accounts",
  "Assets",
  "Cart",
  "Catalog",
  "Emails",
  "Groups",
  "Inventory",
  "MediaRecords",
  "Notifications",
  "Orders",
  "Packages",
  "Products",
  "Revisions",
  "roles",
  "SellerShops",
  "Shipping",
  "Shops",
  "Tags",
  "Templates",
  "Themes",
  "Translations",
  "users"
].forEach((collectionName) => {
  mockContext.collections[collectionName] = {
    insert() {
      throw new Error("insert mongo method is deprecated, use insertOne or insertMany");
    },
    remove() {
      throw new Error("remove mongo method is deprecated, use deleteOne or deleteMany");
    },
    update() {
      throw new Error("update mongo method is deprecated, use updateOne or updateMany");
    },
    deleteOne: jest.fn().mockName(`${collectionName}.deleteOne`).mockReturnValue(Promise.resolve({
      deletedCount: 1
    })),
    deleteMany: jest.fn().mockName(`${collectionName}.deleteMany`),
    find: jest
      .fn()
      .mockName(`${collectionName}.find`)
      .mockReturnThis(),
    findOne: jest.fn().mockName(`${collectionName}.findOne`),
    insertOne: jest.fn().mockName(`${collectionName}.insertOne`),
    insertMany: jest.fn().mockName(`${collectionName}.insertMany`),
    toArray: jest.fn().mockName(`${collectionName}.toArray`),
    updateOne: jest.fn().mockName(`${collectionName}.updateOne`).mockReturnValue(Promise.resolve({
      matchedCount: 1,
      modifiedCount: 1
    })),
    updateMany: jest.fn().mockName(`${collectionName}.updateMany`)
  };
});

mockContext.collections.Media = {
  find: jest.fn().mockName("Media.find"),
  findLocal: jest.fn().mockName("Media.findLocal"),
  findOne: jest.fn().mockName("Media.findOne"),
  findOneLocal: jest.fn().mockName("Media.findOneLocal"),
  insert: jest.fn().mockName("Media.insert"),
  update: jest.fn().mockName("Media.update"),
  remove: jest.fn().mockName("Media.remove")
};

export default mockContext;
