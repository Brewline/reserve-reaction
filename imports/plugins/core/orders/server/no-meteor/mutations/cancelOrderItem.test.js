import cancelOrderItem from "./cancelOrderItem";
import Factory from "/imports/test-utils/helpers/factory";
import mockContext from "/imports/test-utils/helpers/mockContext";

beforeEach(() => {
  jest.resetAllMocks();
});

test("throws if orderId isn't supplied", async () => {
  await expect(cancelOrderItem(mockContext, {
    itemId: "abc",
    cancelQuantity: 1
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if itemId isn't supplied", async () => {
  await expect(cancelOrderItem(mockContext, {
    orderId: "abc",
    cancelQuantity: 1
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if cancelQuantity isn't supplied", async () => {
  await expect(cancelOrderItem(mockContext, {
    itemId: "abc",
    orderId: "abc"
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if cancelQuantity is 0", async () => {
  await expect(cancelOrderItem(mockContext, {
    itemId: "abc",
    orderId: "abc",
    cancelQuantity: 0
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if cancelQuantity is negative", async () => {
  await expect(cancelOrderItem(mockContext, {
    itemId: "abc",
    orderId: "abc",
    cancelQuantity: -1
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if the order doesn't exist", async () => {
  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve(null));

  await expect(cancelOrderItem(mockContext, {
    itemId: "abc",
    orderId: "abc",
    cancelQuantity: 1
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if the order item doesn't exist", async () => {
  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve({
    shipping: [
      {
        items: [
          {
            _id: "xyz",
            workflow: {
              status: "new",
              workflow: ["new"]
            }
          }
        ],
        workflow: {
          status: "new",
          workflow: ["new"]
        }
      }
    ],
    shopId: "SHOP_ID",
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  }));

  mockContext.userHasPermission.mockReturnValueOnce(true);

  await expect(cancelOrderItem(mockContext, {
    itemId: "abc",
    orderId: "abc",
    cancelQuantity: 1
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if permission check fails", async () => {
  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve({
    shipping: [
      {
        items: []
      }
    ],
    shopId: "SHOP_ID"
  }));

  mockContext.userHasPermission.mockReturnValueOnce(false);

  await expect(cancelOrderItem(mockContext, {
    itemId: "abc",
    orderId: "abc",
    cancelQuantity: 1
  })).rejects.toThrowErrorMatchingSnapshot();

  expect(mockContext.userHasPermission).toHaveBeenCalledWith(["orders"], "SHOP_ID");
});

test("throws if user who placed order tries to cancel at invalid current item status", async () => {
  const accountId = "ACCOUNT_ID";

  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve({
    accountId,
    shipping: [
      Factory.OrderFulfillmentGroup.makeOne({
        items: [
          Factory.OrderItem.makeOne({
            _id: "ITEM_1",
            quantity: 1,
            price: {
              amount: 1,
              currencyCode: "USD"
            },
            subtotal: 1,
            workflow: {
              status: "processing",
              workflow: ["new", "processing"]
            }
          })
        ],
        workflow: {
          status: "new",
          workflow: ["new"]
        }
      })
    ],
    shopId: "SHOP_ID",
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  }));

  mockContext.accountId = accountId;

  await expect(cancelOrderItem(mockContext, {
    itemId: "ITEM_1",
    orderId: "ORDER_1",
    cancelQuantity: 1
  })).rejects.toThrowErrorMatchingSnapshot();

  mockContext.accountId = null;
});

test("throws if user who placed order tries to cancel at invalid current order status", async () => {
  const accountId = "ACCOUNT_ID";

  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve({
    accountId,
    shipping: [
      Factory.OrderFulfillmentGroup.makeOne({
        items: [
          Factory.OrderItem.makeOne({
            _id: "ITEM_1",
            quantity: 1,
            price: {
              amount: 1,
              currencyCode: "USD"
            },
            subtotal: 1,
            workflow: {
              status: "new",
              workflow: ["new"]
            }
          })
        ],
        workflow: {
          status: "new",
          workflow: ["new"]
        }
      })
    ],
    shopId: "SHOP_ID",
    workflow: {
      status: "processing",
      workflow: ["new", "processing"]
    }
  }));

  mockContext.accountId = accountId;

  await expect(cancelOrderItem(mockContext, {
    itemId: "ITEM_1",
    orderId: "ORDER_1",
    cancelQuantity: 1
  })).rejects.toThrowErrorMatchingSnapshot();

  mockContext.accountId = null;
});

test("throws if cancelQuantity is greater than item quantity", async () => {
  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve({
    shipping: [
      {
        items: [
          { _id: "abc", quantity: 1 }
        ]
      }
    ],
    shopId: "SHOP_ID"
  }));

  mockContext.userHasPermission.mockReturnValueOnce(true);

  await expect(cancelOrderItem(mockContext, {
    itemId: "abc",
    orderId: "abc",
    cancelQuantity: 2
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if the database update fails", async () => {
  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve({
    shipping: [
      Factory.OrderFulfillmentGroup.makeOne({
        items: [
          Factory.OrderItem.makeOne({
            _id: "abc",
            quantity: 1,
            price: {
              amount: 1,
              currencyCode: "USD"
            },
            subtotal: 1,
            workflow: {
              status: "new",
              workflow: ["new"]
            }
          })
        ],
        workflow: {
          status: "new",
          workflow: ["new"]
        }
      })
    ],
    shopId: "SHOP_ID",
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  }));

  mockContext.userHasPermission.mockReturnValueOnce(true);

  mockContext.collections.Orders.findOneAndUpdate.mockReturnValueOnce(Promise.resolve({
    modifiedCount: 0
  }));

  await expect(cancelOrderItem(mockContext, {
    itemId: "abc",
    orderId: "abc",
    cancelQuantity: 1
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("skips permission check if context.isInternalCall", async () => {
  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve({
    shipping: [
      Factory.OrderFulfillmentGroup.makeOne({
        items: [
          Factory.OrderItem.makeOne({
            _id: "abc",
            quantity: 1,
            price: {
              amount: 1,
              currencyCode: "USD"
            },
            subtotal: 1,
            workflow: {
              status: "new",
              workflow: ["new"]
            }
          })
        ],
        workflow: {
          status: "new",
          workflow: ["new"]
        }
      })
    ],
    shopId: "SHOP_ID",
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  }));

  mockContext.collections.Orders.findOneAndUpdate.mockReturnValueOnce(Promise.resolve({
    modifiedCount: 1,
    value: {}
  }));

  mockContext.isInternalCall = true;

  await cancelOrderItem(mockContext, {
    itemId: "abc",
    orderId: "abc",
    cancelQuantity: 1
  });

  delete mockContext.isInternalCall;

  expect(mockContext.userHasPermission).not.toHaveBeenCalled();
});

test("cancels all of an item", async () => {
  const item1 = Factory.OrderItem.makeOne({
    _id: "ITEM_1",
    cancelReason: null,
    quantity: 1,
    price: {
      amount: 1,
      currencyCode: "USD"
    },
    subtotal: 1,
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  });

  const item2 = Factory.OrderItem.makeOne({
    _id: "ITEM_2",
    cancelReason: null,
    quantity: 1,
    price: {
      amount: 1,
      currencyCode: "USD"
    },
    subtotal: 1,
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  });

  const group = Factory.OrderFulfillmentGroup.makeOne({
    items: [item1, item2]
  });

  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve({
    _id: "ORDER_1",
    shipping: [group],
    shopId: "SHOP_ID",
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  }));

  mockContext.userHasPermission.mockReturnValueOnce(true);

  mockContext.collections.Orders.findOneAndUpdate.mockReturnValueOnce(Promise.resolve({
    modifiedCount: 1,
    value: {}
  }));

  await cancelOrderItem(mockContext, {
    itemId: "ITEM_1",
    orderId: "ORDER_1",
    cancelQuantity: 1,
    reason: "REASON"
  });

  expect(mockContext.collections.Orders.findOneAndUpdate).toHaveBeenCalledWith(
    { _id: "ORDER_1" },
    {
      $set: {
        shipping: [
          {
            ...group,
            items: [
              {
                ...item1,
                cancelReason: "REASON",
                workflow: {
                  status: "coreOrderItemWorkflow/canceled",
                  workflow: ["new", "coreOrderItemWorkflow/canceled"]
                }
              },
              item2
            ]
          }
        ],
        updatedAt: jasmine.any(Date)
      }
    },
    { returnOriginal: false }
  );
});

test("cancels some of an item", async () => {
  const item1 = Factory.OrderItem.makeOne({
    _id: "ITEM_1",
    cancelReason: null,
    price: {
      amount: 5,
      currencyCode: "USD"
    },
    quantity: 5,
    subtotal: 25,
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  });

  const item2 = Factory.OrderItem.makeOne({
    _id: "ITEM_2",
    cancelReason: null,
    price: {
      amount: 5,
      currencyCode: "USD"
    },
    quantity: 1,
    subtotal: 5,
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  });

  const group = Factory.OrderFulfillmentGroup.makeOne({
    items: [item1, item2],
    itemIds: [item1._id, item2._id]
  });

  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve({
    _id: "ORDER_1",
    shipping: [group],
    shopId: "SHOP_ID",
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  }));

  mockContext.userHasPermission.mockReturnValueOnce(true);

  mockContext.collections.Orders.findOneAndUpdate.mockReturnValueOnce(Promise.resolve({
    modifiedCount: 1,
    value: {}
  }));

  await cancelOrderItem(mockContext, {
    itemId: "ITEM_1",
    orderId: "ORDER_1",
    cancelQuantity: 2,
    reason: "REASON"
  });

  expect(mockContext.collections.Orders.findOneAndUpdate).toHaveBeenCalledWith(
    { _id: "ORDER_1" },
    {
      $set: {
        shipping: [
          {
            ...group,
            items: [
              {
                ...item1,
                cancelReason: "REASON",
                quantity: 2,
                subtotal: 10,
                workflow: {
                  status: "coreOrderItemWorkflow/canceled",
                  workflow: ["new", "coreOrderItemWorkflow/canceled"]
                }
              },
              item2,
              {
                ...item1,
                _id: jasmine.any(String),
                quantity: 3,
                subtotal: 15,
                workflow: {
                  status: "new",
                  workflow: ["new"]
                }
              }
            ],
            itemIds: [item1._id, item2._id, jasmine.any(String)]
          }
        ],
        updatedAt: jasmine.any(Date)
      }
    },
    { returnOriginal: false }
  );
});

test("cancels the group if all items are canceled", async () => {
  const item1 = Factory.OrderItem.makeOne({
    _id: "ITEM_1",
    cancelReason: null,
    quantity: 1,
    price: {
      amount: 1,
      currencyCode: "USD"
    },
    subtotal: 1,
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  });

  const item2 = Factory.OrderItem.makeOne({
    _id: "ITEM_2",
    cancelReason: null,
    quantity: 1,
    price: {
      amount: 1,
      currencyCode: "USD"
    },
    subtotal: 1,
    workflow: {
      status: "coreOrderItemWorkflow/canceled",
      workflow: ["new", "coreOrderItemWorkflow/canceled"]
    }
  });

  const group1 = Factory.OrderFulfillmentGroup.makeOne({
    items: [item1, item2],
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  });

  const group2 = Factory.OrderFulfillmentGroup.makeOne({
    items: Factory.OrderItem.makeMany(3, {
      price: {
        amount: 1,
        currencyCode: "USD"
      },
      quantity: 1,
      subtotal: 1,
      workflow: {
        status: "new",
        workflow: ["new"]
      }
    }),
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  });

  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve({
    _id: "ORDER_1",
    shipping: [group1, group2],
    shopId: "SHOP_ID",
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  }));

  mockContext.userHasPermission.mockReturnValueOnce(true);

  mockContext.collections.Orders.findOneAndUpdate.mockReturnValueOnce(Promise.resolve({
    modifiedCount: 1,
    value: {}
  }));

  await cancelOrderItem(mockContext, {
    itemId: "ITEM_1",
    orderId: "ORDER_1",
    cancelQuantity: 1
  });

  expect(mockContext.collections.Orders.findOneAndUpdate).toHaveBeenCalledWith(
    { _id: "ORDER_1" },
    {
      $set: {
        shipping: [
          {
            ...group1,
            items: [
              {
                ...item1,
                cancelReason: null,
                workflow: {
                  status: "coreOrderItemWorkflow/canceled",
                  workflow: ["new", "coreOrderItemWorkflow/canceled"]
                }
              },
              item2
            ],
            workflow: {
              status: "coreOrderWorkflow/canceled",
              workflow: ["new", "coreOrderWorkflow/canceled"]
            }
          },
          group2
        ],
        updatedAt: jasmine.any(Date)
      }
    },
    { returnOriginal: false }
  );
});

test("cancels the order and emits afterOrderCancel if all groups are canceled", async () => {
  const item1 = Factory.OrderItem.makeOne({
    _id: "ITEM_1",
    cancelReason: null,
    quantity: 1,
    price: {
      amount: 1,
      currencyCode: "USD"
    },
    subtotal: 1,
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  });

  const item2 = Factory.OrderItem.makeOne({
    _id: "ITEM_2",
    cancelReason: null,
    quantity: 1,
    price: {
      amount: 1,
      currencyCode: "USD"
    },
    subtotal: 1,
    workflow: {
      status: "coreOrderItemWorkflow/canceled",
      workflow: ["new", "coreOrderItemWorkflow/canceled"]
    }
  });

  const group1 = Factory.OrderFulfillmentGroup.makeOne({
    items: [item1, item2],
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  });

  const group2 = Factory.OrderFulfillmentGroup.makeOne({
    items: Factory.OrderItem.makeMany(3, {
      price: {
        amount: 1,
        currencyCode: "USD"
      },
      quantity: 1,
      subtotal: 1,
      workflow: {
        status: "coreOrderItemWorkflow/canceled",
        workflow: ["new", "coreOrderItemWorkflow/canceled"]
      }
    }),
    workflow: {
      status: "coreOrderWorkflow/canceled",
      workflow: ["new", "coreOrderWorkflow/canceled"]
    }
  });

  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve({
    _id: "ORDER_1",
    shipping: [group1, group2],
    shopId: "SHOP_ID",
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  }));

  mockContext.userHasPermission.mockReturnValueOnce(true);

  mockContext.collections.Orders.findOneAndUpdate.mockReturnValueOnce(Promise.resolve({
    modifiedCount: 1,
    value: {}
  }));

  await cancelOrderItem(mockContext, {
    itemId: "ITEM_1",
    orderId: "ORDER_1",
    cancelQuantity: 1
  });

  expect(mockContext.collections.Orders.findOneAndUpdate).toHaveBeenCalledWith(
    { _id: "ORDER_1" },
    {
      $set: {
        shipping: [
          {
            ...group1,
            items: [
              {
                ...item1,
                cancelReason: null,
                workflow: {
                  status: "coreOrderItemWorkflow/canceled",
                  workflow: ["new", "coreOrderItemWorkflow/canceled"]
                }
              },
              {
                ...item2,
                cancelReason: null,
                workflow: {
                  status: "coreOrderItemWorkflow/canceled",
                  workflow: ["new", "coreOrderItemWorkflow/canceled"]
                }
              }
            ],
            workflow: {
              status: "coreOrderWorkflow/canceled",
              workflow: ["new", "coreOrderWorkflow/canceled"]
            }
          },
          group2
        ],
        updatedAt: jasmine.any(Date),
        workflow: {
          status: "coreOrderWorkflow/canceled",
          workflow: ["new", "coreOrderWorkflow/canceled"]
        }
      }
    },
    { returnOriginal: false }
  );
});

test("succeeds if already canceled, but does not push canceled status again", async () => {
  const item1 = Factory.OrderItem.makeOne({
    _id: "ITEM_1",
    cancelReason: null,
    quantity: 1,
    price: {
      amount: 1,
      currencyCode: "USD"
    },
    subtotal: 1,
    workflow: {
      status: "coreOrderItemWorkflow/canceled",
      workflow: ["new", "coreOrderItemWorkflow/canceled"]
    }
  });

  const item2 = Factory.OrderItem.makeOne({
    _id: "ITEM_2",
    cancelReason: null,
    quantity: 1,
    price: {
      amount: 1,
      currencyCode: "USD"
    },
    subtotal: 1,
    workflow: {
      status: "coreOrderItemWorkflow/canceled",
      workflow: ["new", "coreOrderItemWorkflow/canceled"]
    }
  });

  const group = Factory.OrderFulfillmentGroup.makeOne({
    items: [item1, item2],
    workflow: {
      status: "coreOrderWorkflow/canceled",
      workflow: ["new", "coreOrderWorkflow/canceled"]
    }
  });

  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve({
    _id: "ORDER_1",
    shipping: [group],
    shopId: "SHOP_ID",
    workflow: {
      status: "coreOrderWorkflow/canceled",
      workflow: ["new", "coreOrderWorkflow/canceled"]
    }
  }));

  mockContext.userHasPermission.mockReturnValueOnce(true);

  mockContext.collections.Orders.findOneAndUpdate.mockReturnValueOnce(Promise.resolve({
    modifiedCount: 1,
    value: {}
  }));

  await cancelOrderItem(mockContext, {
    itemId: "ITEM_1",
    orderId: "ORDER_1",
    cancelQuantity: 1,
    reason: "REASON"
  });

  expect(mockContext.collections.Orders.findOneAndUpdate).toHaveBeenCalledWith(
    { _id: "ORDER_1" },
    {
      $set: {
        shipping: [
          {
            ...group,
            items: [
              {
                ...item1,
                cancelReason: "REASON",
                workflow: {
                  status: "coreOrderItemWorkflow/canceled",
                  workflow: ["new", "coreOrderItemWorkflow/canceled"]
                }
              },
              item2
            ]
          }
        ],
        updatedAt: jasmine.any(Date)
      }
    },
    { returnOriginal: false }
  );
});
