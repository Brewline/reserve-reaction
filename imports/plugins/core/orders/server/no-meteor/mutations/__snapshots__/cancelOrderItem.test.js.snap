// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws if cancelQuantity is 0 1`] = `"Cancel quantity must be at least 1"`;

exports[`throws if cancelQuantity is greater than item quantity 1`] = `"cancelQuantity may not be greater than item quantity"`;

exports[`throws if cancelQuantity is negative 1`] = `"Cancel quantity must be at least 1"`;

exports[`throws if cancelQuantity isn't supplied 1`] = `"Cancel quantity is required"`;

exports[`throws if itemId isn't supplied 1`] = `"Item ID is required"`;

exports[`throws if orderId isn't supplied 1`] = `"Order ID is required"`;

exports[`throws if permission check fails 1`] = `"Access Denied"`;

exports[`throws if the database update fails 1`] = `"Unable to update order"`;

exports[`throws if the order doesn't exist 1`] = `"Order not found"`;

exports[`throws if the order item doesn't exist 1`] = `"Order item not found"`;

exports[`throws if user who placed order tries to cancel at invalid current item status 1`] = `"Item status (processing) is not one of: new"`;

exports[`throws if user who placed order tries to cancel at invalid current order status 1`] = `"Order status (processing) is not one of: new"`;
